### PR TITLE
Javalab: Import from backpack

### PR DESCRIPTION
**Pull request by molly-moen**
Update javalab's backpack import menu to load the list of files in the user's backpack and import 1 or more files from their backpack into their current project. We re-load the list of files every time the user clicks the backpack button. This way if they are working in multiple tabs they can save to backpack in one tab and import into the other without needing to reload the page.

### No files checked: import button is disabled
Screen Shot 2021-07-29 at 9 00 43 AM
Screen Shot 2021-07-29 at 9 02 36 AM

### Some files checked: import button enabled
Screen Shot 2021-07-29 at 9 02 46 AM

Screen Shot 2021-07-29 at 9 00 52 AM

### Message if they have no files in their backpack
Screen Shot 2021-07-29 at 8 52 02 AM
Screen Shot 2021-07-29 at 8 52 10 AM

### Message if we hit an error loading files
Screen Shot 2021-07-29 at 9 05 02 AM

## Links

- spec: [Backpack](https://docs.google.com/document/d/1Y01X2_rrVt8f4y8yHpm3BBpq31Ra9CSkOLYycwQ5or8/edit#bookmark=id.fwf06t82gl9u)
- jira ticket: [CSA-524](https://codedotorg.atlassian.net/browse/CSA-524)
- [figma](https://www.figma.com/file/fZo4fcZmDUioOCaz04Gza4/Javalab-Design-Spec?node-id=0%3A1)

## Testing story
Tested locally on Firefox and Chrome.

## Follow-up work
Currently we allow any file to be imported with no errors/warnings. We have a [follow up task](https://codedotorg.atlassian.net/browse/CSA-527) to give a warning when a user is overwriting a visible file, and an error if they are overwriting a support file. 

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/i18n/javalab/en_us.json [hunk 1]
```diff
@@ -34,5 +34,7 @@
   "saveToBackpack": "Save to Backpack",
   "twoMainMethods": "Error: Your code can only have one main method. We found at least two classes with main methods.",
   "backpackSaveInProgress": "Error: Save to Backpack is currently in progress, cannot start another Save to Backpack now.",
-  "unknownError": "Oops. We hit an error on our side. Try running your program again. If this error continues, contact us at support@code.org. Be sure to include the connection ID {connectionId} in your message.\n{type}"
+  "unknownError": "Oops. We hit an error on our side. Try running your program again. If this error continues, contact us at support@code.org. Be sure to include the connection ID {connectionId} in your message.\n{type}",
+  "backpackListLoadError": "An error occurred loading your backpack. Try again.",
+  "emptyBackpackMessage": "Files saved to your backpack will appear here."
```
Comment (jmkulwik) [inline apps/i18n/javalab/en_us.json]: NIT: I may be wrong here, but I think we generally try to keep these files alphabetized by key. i.e. "emptyBackpackMessage" would go under "editor"

**Review by jmkulwik — COMMENTED** [review]
If you get the time, this change seems like a good candidate for some unit testing.

File: apps/i18n/javalab/en_us.json (hunk 1 above)
Comment (molly-moen) [reply apps/i18n/javalab/en_us.json]: yep, will fix

File: apps/src/javalab/Backpack.jsx [hunk 2]
```diff
@@ -76,45 +152,62 @@ export class Backpack extends Component {
...
+                    <label
+                      htmlFor={`backpack-file-${index}`}
+                      style={styles.fileListLabel}
+                    >
+                      {filename}
+                    </label>
+                  </div>
+                ))}
+                <JavalabButton
+                  text="Import"
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/Backpack.jsx]: nit: localized string for this?

File: apps/src/javalab/JavalabEditor.jsx [hunk 3]
```diff
@@ -410,6 +412,30 @@ class JavalabEditor extends React.Component {
...
+      for (const key in fileMetadata) {
+        if (fileMetadata[key] === filename) {
+          editorKey = key;
+        }
+      }
+      const editor = this.editors[editorKey];
+      editor.dispatch({
+        changes: {from: 0, to: editor.state.doc.length, insert: fileContents}
+      });
+      this.props.setSource(filename, fileContents);
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabEditor.jsx:431]: Curious - does `setSource` not overwrite the editor automatically?

File: apps/src/javalab/Backpack.jsx [hunk 4]
```diff
@@ -5,29 +5,59 @@ import color from '@cdo/apps/util/color';
...
+      selectedFiles: [],
+      backpackFilenames: []
+    });
+    if (this.props.backpackApi.hasBackpack()) {
+      this.setState({backpackFilesLoading: true});
+      this.props.backpackApi.getFileList(
+        this.onFileListLoadError,
+        this.onFileListLoadSuccess
+      );
+    }
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/Backpack.jsx:45]: What do users see if the backpack API isn't available? would it just be empty?

**Review by sanchitmalhotra126 — APPROVED** [review]
Couple small things, otherwise LGTM!

File: apps/src/javalab/Backpack.jsx (hunk 4 above)
Comment (molly-moen) [reply apps/src/javalab/Backpack.jsx:45]: if the api has a load error they see the error message. If they haven't added anything to the backpack yet/we got a null channel id from the backend they see you have no files.

File: apps/src/javalab/JavalabEditor.jsx (hunk 3 above)
Comment (molly-moen) [reply apps/src/javalab/JavalabEditor.jsx:431]: It doesn't. We have sort of a weird setup that could be looked at again post-pilot. The sources populate the editor, but then the editor controls its own state going forward (via codemirror) and sends updates back to sources for saving purposes. Before now we never put new contents into an existing editor after load.

File: apps/src/javalab/Backpack.jsx (hunk 2 above)
Comment (molly-moen) [reply apps/src/javalab/Backpack.jsx]: done!

**Comment by molly-moen** [comment]
@jmkulwik added some unit tests!

**Review by jmkulwik — APPROVED** [review]
Nice work! LGTM!

**Pull request merged**